lib/tinyusb: Update to version 0.19.0-24.

dpgeorge opened the pull request:
### Summary

This PR updates TinyUSB to 0.19.0-24.  That's 0.19.0 plus the following changes:
- remove obsolete dcd_esp32sx
- fix for HID stylus descriptor and HID example
- typos, docs and generator scripts
- MTP fix
- DWC2 enumeration when EP0 size=8
- DWC2 fix for EP0 IN
- stm32 FSDEV IRQ remapping fix
- DWC2 ZLP fix

The reason we need the extra 24 commits is due to a bug with TinyUSB's handling of zero-length-packets in the DWC2 (Synopsis) backend, which affects the stm32 port.  That's fixed by https://github.com/hathach/tinyusb/pull/3293

(The same DWC2 bug affects ESP32-Sx, but needs to be fixed separately because we are using the IDF component manager to get TinyUSB for the esp32 port.)

### Testing

Tested on Linux and macOS using the test from #18402, with the following boards:
- rp2 Pico: ok
- rp2 Pico 2 RISCV: ok
- mimxrt TEENSY40: ok (although fails DATA IN for largest packet, but that's also on current TinyUSB)
- samd ADAFRUIT_ITSYBITSY_M0_EXPRESS: ok (but fails DATA OUT verified for 16k packets, but that's also on current TinyUSB)
- stm32 PYBV10: ok
- stm32 OPENMV_N6: ok (but known issue with soft reset)
- nrf ARDUINO_NANO_33_BLE_SENSE: ok

### Trade-offs and Alternatives

Not much alternative here, we anyway need to update TinyUSB to keep up with the latest version, and get STM32N6 support.

dpgeorge committed:
lib/tinyusb: Update to version 0.19.0-24.

This is 0.19.0 plus the following changes:
- remove obsolete dcd_esp32sx
- fix for HID stylus descriptor and HID example
- typos, docs and generator scripts
- MTP fix
- DWC2 enumeration when EP0 size=8
- DWC2 fix for EP0 IN
- stm32 FSDEV IRQ remapping fix
- DWC2 ZLP fix

The reason we need the extra 24 commits is due to a bug with TinyUSB's
handling of zero-length-packets in the DWC2 (Synopsis) backend, which
affects the stm32 port.  That's fixed by
https://github.com/hathach/tinyusb/pull/3293

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
nrf/drivers/usb: Provide macros for nrf errata.

Signed-off-by: Damien George <damien@micropython.org>

projectgus reviewed (approved):
Change seems good to me, it's clearly necessary for the DWC2 ZLP fix at least.

Code size impact is larger than we'd like, but again unsure if there's much alternative.

One thing to note is that this update will bring in the header change from 3ec2e367decd8da968b1b85b65264bbab367bc9c - as we now expect this to be enabled universally then we could take out the `#if` guard around it.

iabdalkader commented:
I've tested this with AE3 and N6 (my DFU bootloader, not the new stm32 stack support), and DFU and CDC seem to work fine.

dpgeorge committed:
shared/tinyusb: Remove macro guard for tx_overwritabe_if_not_connected.

It's now available in the version of TinyUSB used by this repository.

Also, update to the newer `tud_cdc_configure_t` struct name and newer
`tud_cdc_configure()` function name.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> The initial delayed enumeration disappeared after I changed the USB hub on my desk

FWIW, I'm using a USB port directly into the (Arch Linux) PC, so the only USB hubs are the internal ones.

And my testing shows that the issue is clearly introduced by the TinyUSB commit I link to above.

dpgeorge commented:
I've rolled this back to just update TinyUSB to 0.19.0-24.  That seems the safest option for the next MicroPython release.

Can work out what the issue is with TinyUSB 0.20.0 at a later date (and it could well be just a MicroPython integration issue, not a problem with TinyUSB itself).

iabdalkader commented:
> I tested `ARDUINO_NANO_33_BLE_SENSE` REV2 again with latest master and it works OK.

Thanks! This confirms that my board is broken.
